### ISSUE TITLE
Questions on implementation

### ISSUE THREAD

**Issue by tomouellette** [instanseg/instanseg#17]
Hi there @alanocallaghan and @ThibautGoldsborough,

I have some interest in porting the model to some different languages (e.g. Rust) but I found the code base a bit cluttered as of now. Before attempting a port to another language, I was just wondering if you could verify this minimal implementation I made ( based on what I read in this repository). Basically, I am just looking to validate that the pre/post-processing is correct.  I've tested it on some images and it seems to reproduce for all tested examples. Thanks and nice model!

**A minimal implementation**

```python
import os
import sys
import torch
import subprocess
import numpy as np
from PIL import Image
from typing import Union
from torchvision.transforms import InterpolationMode, Resize

class InstansegRunner:
    """Apply the instanseg model on brightfield images

    Parameters
    ----------
    model_path : str
        Path to Instanseg torchscript model
    device : str
        Valid torch device ("cpu", "cuda", "mps")
    normalize_percentile : float
        Percentile for contrast stretching of input image
    normalize_epsilon : float
        Small float to avoid division by zero errors during normalization
    interpolation : Union[int, InterpolationMode]
        Type of kernel for resizing mask after segmenting

    References
    ----------
    1. Goldsborough et al. "InstanSeg: an embedding-based instance segmentation 
       algorithm optimized for accurate, efficient and portable cell segmentation". 
       arXiv (2024). https://arxiv.org/abs/2408.15954.
    2. https://github.com/instanseg/instanseg: Original model implementation
    """
    
    _base_mpp  = 0.5
    _base_shape = (256, 256)
    
    def __init__(
        self,
        model_path: str,       
        device: str = "cpu",
        normalize_percentile: float = 0.1,
        normalize_epsilon: float = 1e-3,
        interpolation: Union[int, InterpolationMode] = InterpolationMode.NEAREST,
    ):          
        if os.path.exists(model_path):
            self.model = torch.jit.load(model_path).to(device)            

        self.device = device
        self.percentile = [normalize_percentile, 100 - normalize_percentile]
        self.eps = normalize_epsilon
        self.interpolation = interpolation
        
    def segment(self, image: Union[str, np.ndarray], microns_per_pixel: float) -> np.ndarray:
        """Segment a 3-channel brightfield image

        Parameters
        ----------
        image : Union[str, np.ndarray]
            Path to image or a numpy array with shape (H, W, 3) or (3, H, W)
        microns_per_pixel : float
            Microns per pixel

        Returns
        -------
        np.ndarray
            A integer-level mask of shape (H, W)
        """
            
        if isinstance(image, str):
            if os.path.exists(image):
                image = np.array(Image.open(image))

        if np.min(image.shape) != 3 or image.ndim != 3:
            raise ValueError("Image must be (3, H, W) or (H, W, 3) image.")
            
        channel_index = np.argmin(image.shape)
        if channel_index == 0:
            image = image.transpose(1, 2, 0)

        image = image.astype(np.float32).copy()
        
        for c in range(3):
            min_i, max_i = np.percentile(image[:, :, c], self.percentile)
            image[:, :, c] = (image[:, :, c] - min_i) / max(self.eps, max_i - min_i)

        image = torch.tensor(image, dtype=torch.float32).permute(2, 0, 1).to(self.device)

        oh, ow = image.shape[-2:]
        if microns_per_pixel != self._base_mpp:
            scale = microns_per_pixel / self._base_mpp
            h, w = int(oh * scale), int(ow * scale)            
        
        image = Resize(size=(h, w), antialias=True)(image)
        image = image.unsqueeze(0) if image.dim() == 2 else image
        
        pad_size = ((np.array(self._base_shape) - np.array([h, w])) / 2).astype(int) + 3
        pad_size = np.minimum(pad_size, [h, w]).min() - 1
        if pad_size > 0:
            image = torch.nn.functional.pad(
                image, 
                (pad_size, pad_size, pad_size, pad_size), 
                mode='constant',
                value=image.max()
            )

        mask = self.model(image.unsqueeze(0))        
        mask = mask[0, 0:1, pad_size:image.shape[-2]-pad_size, pad_size:image.shape[-1]-pad_size]
        mask = Resize((h, w), interpolation=self.interpolation)(mask)[0]
        
        return mask.cpu().detach().numpy()
```

**Comment by ThibautGoldsborough** [comment]
Hi @tomouellette. Thanks for your suggestions. 

Don't worry about the pad size / padding. InstanSeg doesn't need a (256,256) input, it should take any (h,w) input.

Do you intend to run InstanSeg on small (fits on the GPU) images? If so your implementation looks correct. (The numpy percentile normalization looks slowish, torch has a torch.quantile function that can be done on the gpu). If you intend to tile images, then it's better to normalize the whole image rather than each tile independently. (A word of warning, stitching labelled outputs is difficult.)

I'd be interested to see if there's a large speedup in rust. In python/java, the bottleneck is the speed of your gpu (~100% utilization) rather than the cpu overhead.

**Comment by tomouellette** [comment]
Thanks for the quick response! So my initial experiments only showed good segmentations when applying the padding and rescaling transform at (256, 256). When I get some time I will run them and post them here to see if there's an optimization.

However, in your `inference.py` code you specify a rescaling at line 125 that takes in the argument `rescaled_pixel_size` ([Line 125](https://github.com/instanseg/instanseg/blob/main/InstanSeg/scripts/inference.py#L125).) The `rescaled_pixel_size` is the one stored in the torchscript model which is 256. Am I missing a particular portion of the code for this because it seems the base pixel size stored in the torchscript model is the hard-coded default?

**Comment by ThibautGoldsborough** [comment]
I'm not sure what `rescaled_pixel_size` you are referring to is. By pixel size we mean the length in microns of one pixel. This value is stored in the torchscript (instanseg.pixel_size) and is 0.5 microns/pixel for the current public models. Your image will have some other pixel size (usually in the range 0.1 - 0.5 microns/pixel), which is usually present in the image metadata. (Sometimes that information is deleted or lost and you'll have to manually set the pixel size.)

**Comment by tomouellette** [comment]
Thanks for quick response again and sorry for the confusion, I must of been thinking of two things at once!

I meant to say line 707 in augmentations ([Line 707](https://github.com/instanseg/instanseg/blob/b9c9d6b6860f548a4b6bb9151270e390d9bfe52e/InstanSeg/utils/augmentations.py#L707)) which uses `self.shape` that is is hardcoded to 256 for padding. It doesn't seem that the augmentation class in `inference.py` is updated with any shape either?

```python
while self.shape is not None and np.any(np.array(resized_data[0].shape) < self.shape[0]):
            pad = int((self.shape[0] - min(resized_data[0].shape)) / 2) + 3
            pad = torch.Tensor([pad, resized_data.shape[1], resized_data.shape[2]]).min().int() - 1
```

And given,

```
scale = (current_pixel_size / requested_pixel_size)
shape = (torch.tensor(image.shape[-2:]) * scale).int().tolist()
```

Looks like padding to a final size of 256 is default for many smaller images or images where scale leads to a resizing below 256?

**Comment by ThibautGoldsborough** [comment]
Ah indeed that is unintended, thank you!. The 256 requirement is for training only not for inference.

`while self.shape is not None and np.any(np.array(resized_data[0].shape) < self.shape[0]) and crop is True:`

should be a quick fix. Thanks for pointing that out.

**Comment by tomouellette** [comment]
I think that's all my questions then - thanks for the clarifications. Ill close the issue (and let you know when I get a chance to implement a version in rust). Cheers!
